File: evals/refs/uf.zig
const std = @import("std");

const UF = struct {
    parent: []usize,
    size: []usize,

    fn find(self: *UF, x: usize) usize {
        var cur = x;
        while (self.parent[cur] != cur) cur = self.parent[cur];
        var i = x;
        while (self.parent[i] != cur) {
            const nxt = self.parent[i];
            self.parent[i] = cur;
            i = nxt;
        }
        return cur;
    }

    fn unite(self: *UF, a: usize, b: usize) void {
        const ra = self.find(a);
        const rb = self.find(b);
        if (ra == rb) return;
        if (self.size[ra] < self.size[rb]) {
            self.parent[ra] = rb;
            self.size[rb] += self.size[ra];
        } else {
            self.parent[rb] = ra;
            self.size[ra] += self.size[rb];
        }
    }
};

pub fn main(init: std.process.Init) !void {
    const arena = init.arena.allocator();

    var in_buf: [4096]u8 = undefined;
    var fr = std.Io.File.stdin().reader(init.io, &in_buf);
    const r = &fr.interface;

    var data: std.ArrayList(u8) = .empty;
    try r.appendRemainingUnlimited(arena, &data);

    var out_buf: [4096]u8 = undefined;
    var fw = std.Io.File.stdout().writer(init.io, &out_buf);
    const w = &fw.interface;

    var it = std.mem.tokenizeAny(u8, data.items, " \n\r\t");
    const n = try std.fmt.parseInt(usize, it.next().?, 10);
    const q = try std.fmt.parseInt(usize, it.next().?, 10);

    var uf: UF = .{
        .parent = try arena.alloc(usize, n),
        .size = try arena.alloc(usize, n),
    };
    var i: usize = 0;
    while (i < n) : (i += 1) {
        uf.parent[i] = i;
        uf.size[i] = 1;
    }

    var k: usize = 0;
    while (k < q) : (k += 1) {
        const op = it.next().?;
        const a = try std.fmt.parseInt(usize, it.next().?, 10);
        const b = try std.fmt.parseInt(usize, it.next().?, 10);
        if (op[0] == 'u') {
            uf.unite(a, b);
        } else {
            const same = uf.find(a) == uf.find(b);
            try w.writeAll(if (same) "true\n" else "false\n");
        }
    }
    try w.flush();
}

File: evals/refs/uf.rs
use std::io::{self, Read, Write};

fn find(parent: &mut Vec<usize>, x: usize) -> usize {
    let mut root = x;
    while parent[root] != root {
        root = parent[root];
    }
    let mut cur = x;
    while parent[cur] != root {
        let nxt = parent[cur];
        parent[cur] = root;
        cur = nxt;
    }
    root
}

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut it = input.split_ascii_whitespace();
    let stdout = io::stdout();
    let mut out = io::BufWriter::new(stdout.lock());

    let n: usize = it.next().unwrap().parse().unwrap();
    let q: usize = it.next().unwrap().parse().unwrap();
    let mut parent: Vec<usize> = (0..n).collect();
    let mut size = vec![1usize; n];

    for _ in 0..q {
        let op = it.next().unwrap();
        let a: usize = it.next().unwrap().parse().unwrap();
        let b: usize = it.next().unwrap().parse().unwrap();
        if op == "u" {
            let ra = find(&mut parent, a);
            let rb = find(&mut parent, b);
            if ra != rb {
                let (big, small) = if size[ra] < size[rb] { (rb, ra) } else { (ra, rb) };
                parent[small] = big;
                size[big] += size[small];
            }
        } else {
            let ra = find(&mut parent, a);
            let rb = find(&mut parent, b);
            out.write_all(if ra == rb { b"true\n" } else { b"false\n" }).unwrap();
        }
    }
}
